guard/src/rules/eval_context.rs

```diff
@@ -1132,23 +1207,154 @@ impl<'value, 'loc: 'value> EvalContext<'value, 'loc> for RootScope<'value, 'loc>
         self.scope
             .resolved_variables
             .insert(variable_name, result.clone());
-        return Ok(result);
+        Ok(result)
     }
 
     fn add_variable_capture_key(
         &mut self,
         variable_name: &'value str,
-        key: &'value PathAwareValue,
+        key: Rc<PathAwareValue>,
     ) -> Result<()> {
         self.scope
             .resolved_variables
             .entry(variable_name)
             .or_default()
-            .push(QueryResult::Resolved(key));
+            .push(QueryResult::Resolved(Rc::clone(&key)));
         Ok(())
     }
 }
 
+pub(crate) fn validate_number_of_params(name: &str, num_args: usize) -> Result<()> {
+    match name {
+        "join" => {
+            if num_args != 2 {
+                return Err(Error::ParseError(format!(
+                    "join function requires 2 arguments be passed, but received {num_args}"
```

Comment:
There probably isn't, but if there's an automated way where we could do this in rust we can re-use the same error message multiple times either dynamically OR by using maybe a `Hashmap` for `<function_name, number_of_expected_params>` we should use it.

Reply:
I changed this up a bit, I think the new implementation is cleaner. 

If we decide to actually abstract the functions into structs I thin this can be handled even cleaner.